File: macos-swiftui/Sources/Core/ArchiveInfo.swift
/* info.json, comments and the transcript.
 *
 * Read ON DEMAND, when a video's page is opened -- never during the scan.
 * info.json carries the full comment tree and routinely runs to several
 * megabytes; parsing every one of them to draw a grid of thumbnails would make
 * opening the app cost what opening every video costs.
 *
 * Both parsers here fail QUIETLY when they are wrong -- a mis-threaded comment
 * section still renders, and a transcript with the rolling duplication left in
 * still looks like a transcript. Neither throws. That is why both are pinned
 * by tests rather than trusted.
 */

import Foundation

// MARK: - info.json

struct VideoInfo {
    let root: [String: Any]

    /// nil when there is no readable info.json -- an ordinary state the layout
    /// contract names, not an error.
    static func load(forEntryDir dir: String) -> VideoInfo? {
        let metaDir = Paths.join(dir, "Video metadata")
        guard let path = ArchiveIndex.infoJSONPath(inMetaDir: metaDir),
              let obj = JSONFile.object(at: path) else { return nil }
        return VideoInfo(root: obj)
    }

    func string(_ key: String) -> String? { root.str(key) }
    func int(_ key: String) -> Int64 { root.int(key) }

    /// The threaded comments. Empty when the video was fetched with
    /// --no-comments, which is ordinary.
    var comments: [Comment] {
        Comment.thread(root.objects("comments"))
    }

    /// Up to 40 tags, joined for display.
    var tags: String {
        root.strings("tags").prefix(40).joined(separator: ", ")
    }
}

// MARK: - Comments

final class Comment: Identifiable {
    /* yt-dlp's own comment id, used to thread replies onto parents. NOT this
     * class's Identifiable id: it is "" whenever the extractor did not send
     * one, and a ForEach over rows with duplicate ids silently drops all but
     * one of them. */
    let commentID: String
    let text: String
    let author: String
    let authorID: String?
    let timeText: String?
    /// -1 when absent.
    let timestamp: Int64
    /// -1 when absent.
    let likeCount: Int64
    let isFavorited: Bool
    let authorIsUploader: Bool
    let isPinned: Bool
    var replies: [Comment] = []

    /// Identity is the object, which is unique by construction.
    var id: ObjectIdentifier { ObjectIdentifier(self) }

    init(from o: [String: Any]) {
        commentID = o.str("id") ?? ""
        text = o.str("text") ?? ""
        author = o.str("author") ?? "(unknown)"
        authorID = o.str("author_id")
        timeText = o.str("_time_text")
        timestamp = o.int("timestamp", default: -1)
        likeCount = o.int("like_count", default: -1)
        isFavorited = o.bool("is_favorited")
        authorIsUploader = o.bool("author_is_uploader")
        isPinned = o.bool("is_pinned")
    }

    /* Thread yt-dlp's flat comment list.
     *
     * yt-dlp writes comments as a FLAT array with a `parent` field that is
     * either "root" or the parent's id, IN NO GUARANTEED ORDER -- a reply can
     * appear before its parent, so this is a two-pass job rather than a fold.
     *
     * Top level is pinned first, then most-liked: the same order YouTube itself
     * shows, which matters because a transcript of a comment section in
     * arbitrary order is a different document from the one people read. */
    static func thread(_ raw: [[String: Any]]) -> [Comment] {
        var tops: [Comment] = []
        var index: [String: Int] = [:]      // id -> position in tops
        var orphans: [(parent: String, comment: Comment)] = []

        for o in raw {
            let parent = o.str("parent")
            let c = Comment(from: o)

            if parent == nil || parent == "root" {
                index[c.commentID] = tops.count
                tops.append(c)
            } else {
                orphans.append((parent!, c))
            }
        }

        for (parent, c) in orphans {
            var slot = index[parent]
            if slot == nil, let dot = parent.firstIndex(of: ".") {
                /* yt-dlp's reply ids are "<parent>.<reply>", so the parent id is
                 * recoverable even when the parent field itself is unhelpful. */
                slot = index[String(parent[parent.startIndex..<dot])]
            }

            if let slot {
                tops[slot].replies.append(c)
            } else {
                /* A reply whose parent is genuinely absent -- a deleted comment,
                 * or a truncated fetch -- is shown at top level rather than
                 * dropped. Silently losing archived text would be the worse
                 * failure. */
                tops.append(c)
            }
        }

        for c in tops {
            c.replies.sort { a, b in
                (a.timestamp < 0 ? 0 : a.timestamp) < (b.timestamp < 0 ? 0 : b.timestamp)
            }
        }

        /* Pinned first, then most-liked, with an explicit tie-break on the
         * original position. Swift's sort is NOT stable, and without that
         * tie-break the order of equally-liked comments -- which is most of a
         * long thread -- would differ between two reads of the same file. */
        return tops.enumerated()
            .sorted { a, b in
                if a.element.isPinned != b.element.isPinned { return a.element.isPinned }
                let al = max(a.element.likeCount, 0)
                let bl = max(b.element.likeCount, 0)
                if al != bl { return al > bl }
                return a.offset < b.offset
            }
            .map { $0.element }
    }

    /// This comment plus every reply under it.
    static func totalCount(_ tops: [Comment]) -> Int {
        tops.reduce(0) { $0 + 1 + $1.replies.count }
    }
}

// MARK: - Transcript

struct Cue: Identifiable {
    let start: Double
    var end: Double
    let text: String

    var id: String { "\(start)-\(text.prefix(24))" }
}

enum Transcript {
    /* A .vtt or .srt turned into a readable transcript.
     *
     * YouTube's auto-generated VTT is a ROLLING TWO-LINE DISPLAY: nearly every
     * cue repeats the previous cue's last line, and words carry inline karaoke
     * timestamps. Read as-is it is unusable as prose, so tags are stripped and
     * the repetition is collapsed. */
    static func cues(atPath path: String) -> [Cue] {
        guard let data = FileManager.default.contents(atPath: path),
              let raw = String(data: data, encoding: .utf8) ?? String(data: data, encoding: .isoLatin1)
        else { return [] }

        /* Normalise line endings first: a .vtt written on Windows would
         * otherwise never match the blank-line block separator. */
        let normalised = raw
            .replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")

        var parsed: [Cue] = []
        for block in normalised.components(separatedBy: "\n\n") {
            let lines = block
                .components(separatedBy: "\n")
                .filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
            if lines.isEmpty { continue }

            guard let timeIdx = lines.firstIndex(where: { $0.contains("-->") }) else { continue }
            let timeline = lines[timeIdx]
            guard let arrow = timeline.range(of: "-->") else { continue }

            guard let start = seconds(String(timeline[timeline.startIndex..<arrow.lowerBound])) else {
                continue
            }
            let end = seconds(String(timeline[arrow.upperBound...])) ?? (start + 3.0)

            let body = lines[(timeIdx + 1)...].joined(separator: " ")
            let text = collapseWhitespace(unescapeEntities(stripInlineTags(body)))
            if text.isEmpty { continue }

            parsed.append(Cue(start: start, end: end, text: text))
        }

        return collapseRolling(parsed)
    }

    /* The rolling-display collapse.
     *
     * The comparison is against the previous cue's FULL text, not against what
     * was last emitted. That distinction is the whole algorithm:
     *
     *   cue 1  "the quick brown fox"
     *   cue 2  "the quick brown fox jumps over"
     *   cue 3  "the quick brown fox jumps over the lazy dog"
     *
     * Emitting the tail of cue 2 puts "jumps over" in the output. Comparing cue
     * 3 against THAT finds no common prefix, so cue 3 is emitted whole and the
     * duplication the collapse exists to remove comes straight back on the
     * third line. Keeping lastFull separate from the output is what makes the
     * third line "the lazy dog".
     *
     * The same defect was in orchid-cobalt's src-tauri/src/archive.rs. It works
     * for two lines and then silently stops, which is why the test in this
     * repo's suite is three lines deep and not two. */
    private static func collapseRolling(_ cues: [Cue]) -> [Cue] {
        var out: [Cue] = []
        var lastFull: String?

        for cue in cues {
            if let previous = lastFull, !out.isEmpty {
                if cue.text == previous {
                    out[out.count - 1].end = max(out[out.count - 1].end, cue.end)
                    continue
                }
                /* The 12-character floor keeps a genuinely repeated short line
                 * ("Yeah." then "Yeah. Right.") from being chopped into
                 * fragments. */
                if cue.text.hasPrefix(previous), previous.count > 12 {
                    let tail = String(cue.text.dropFirst(previous.count))
                        .trimmingCharacters(in: .whitespaces)
                    if tail.isEmpty {
                        out[out.count - 1].end = max(out[out.count - 1].end, cue.end)
                    } else {
                        out.append(Cue(start: cue.start, end: cue.end, text: tail))
                    }
                    lastFull = cue.text
                    continue
                }
            }

            out.append(cue)
            lastFull = cue.text
        }
        return out
    }

    /* [hh:]mm:ss[.,]mmm -- the leading run of digits, colons and a decimal
     * mark. Anything after it (VTT cue settings such as "align:start
     * position:0%") is ignored. */
    static func seconds(_ text: String) -> Double? {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        var head = ""
        for ch in trimmed {
            if ch.isNumber || ch == ":" || ch == "." || ch == "," {
                head.append(ch)
            } else {
                break
            }
        }
        if head.isEmpty { return nil }

        var whole = head
        var fraction = ""
        if let mark = head.firstIndex(where: { $0 == "." || $0 == "," }) {
            whole = String(head[head.startIndex..<mark])
            fraction = String(head[head.index(after: mark)...])
        }

        let parts = whole.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
        var h = 0.0, m = 0.0, s = 0.0
        if parts.count == 3 {
            h = Double(parts[0]) ?? 0
            m = Double(parts[1]) ?? 0
            s = Double(parts[2]) ?? 0
        } else if parts.count == 2 {
            m = Double(parts[0]) ?? 0
            s = Double(parts[1]) ?? 0
        } else {
            return nil
        }

        /* ".5" is 500ms, not 5ms -- pad on the RIGHT to three digits. */
        var ms = 0.0
        if !fraction.isEmpty {
            let padded = String((fraction + "000").prefix(3))
            ms = Double(padded) ?? 0
        }

        return h * 3600 + m * 60 + s + ms / 1000
    }

    /* Removes <c>, </c>, <v Name>, and the per-word <00:00:01.234> karaoke
     * timestamps YouTube's ASR emits. A hand-rolled scanner rather than a
     * regular expression: the grammar is "everything between < and >", and
     * NSRegularExpression on a multi-megabyte transcript is not free. */
    static func stripInlineTags(_ s: String) -> String {
        var out = ""
        out.reserveCapacity(s.count)
        var depth = 0
        for ch in s {
            if ch == "<" {
                depth += 1
            } else if ch == ">" {
                if depth > 0 { depth -= 1 }
            } else if depth == 0 {
                out.append(ch)
            }
        }
        return out
    }

    static func unescapeEntities(_ s: String) -> String {
        var out = s
        let pairs = [
            ("&amp;", "&"), ("&lt;", "<"), ("&gt;", ">"),
            ("&quot;", "\""), ("&#39;", "'"), ("&nbsp;", " "),
        ]
        for (from, to) in pairs {
            out = out.replacingOccurrences(of: from, with: to)
        }
        return out
    }

    static func collapseWhitespace(_ s: String) -> String {
        s.split(whereSeparator: { $0 == " " || $0 == "\t" || $0 == "\n" || $0 == "\r" })
            .joined(separator: " ")
    }

    /* The filenames cannot tell an auto-generated track from a human-written
     * one -- --write-subs and --write-auto-subs both land in Subtitles/ under
     * the same base name. The CONTENTS can: ASR output carries per-word karaoke
     * tags and cue-positioning directives that uploaded tracks do not.
     *
     * Only the head is read, and these files can be large. */
    static func isAutoGenerated(atPath path: String) -> Bool {
        guard let handle = FileHandle(forReadingAtPath: path) else { return false }
        defer { try? handle.close() }
        let data = handle.readData(ofLength: 8000)
        guard let head = String(data: data, encoding: .utf8)
            ?? String(data: data, encoding: .isoLatin1) else { return false }

        return head.contains("<c.") || head.contains("<c>")
            || head.contains("align:start position:")
    }
}

File: windows-winui/YtdlWin/Core/ArchiveInfo.cs
/* info.json, comments and the transcript.
 *
 * Read ON DEMAND, when a video's page is opened -- never during the scan.
 * info.json carries the full comment tree and routinely runs to several
 * megabytes; parsing every one of them to draw a grid of thumbnails would make
 * opening the app cost what opening every video costs.
 *
 * Both parsers here fail QUIETLY when they are wrong -- a mis-threaded comment
 * section still renders, and a transcript with the rolling duplication left in
 * still looks like a transcript. Neither throws. That is why both are pinned by
 * tests rather than trusted.
 */

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;

namespace YtdlWin.Core;

// MARK: - info.json

public sealed class VideoInfo
{
    public JsonElement Root { get; }

    private VideoInfo(JsonElement root) => Root = root;

    /// null when there is no readable info.json -- an ordinary state the layout
    /// contract names, not an error.
    public static VideoInfo? Load(string entryDir)
    {
        var metaDir = Paths.Join(entryDir, "Video metadata");
        var path = ArchiveIndex.InfoJsonPath(metaDir);
        if (path is null) return null;
        var obj = JsonFile.Object(path);
        return obj is null ? null : new VideoInfo(obj.Value);
    }

    public string? String(string key) => Root.Str(key);
    public long Int(string key) => Root.Int(key);

    /// The threaded comments. Empty when the video was fetched with
    /// --no-comments, which is ordinary.
    public List<Comment> Comments => Comment.Thread(Root.Objects("comments"));

    /// Up to 40 tags, joined for display.
    public string Tags => string.Join(", ", Root.Strings("tags").Take(40));
}

// MARK: - Comments

public sealed class Comment
{
    /* yt-dlp's own comment id, used to thread replies onto parents. Deliberately
     * not treated as an identity anywhere: it is "" whenever the extractor did
     * not send one, and a list keyed on it would collapse every such comment
     * into one row. */
    public string CommentId { get; }
    public string Text { get; }
    public string Author { get; }
    public string? AuthorId { get; }
    public string? TimeText { get; }
    /// -1 when absent.
    public long Timestamp { get; }
    /// -1 when absent.
    public long LikeCount { get; }
    public bool IsFavorited { get; }
    public bool AuthorIsUploader { get; }
    public bool IsPinned { get; }
    public List<Comment> Replies { get; } = new();

    public Comment(JsonElement o)
    {
        CommentId = o.Str("id") ?? "";
        Text = o.Str("text") ?? "";
        Author = o.Str("author") ?? "(unknown)";
        AuthorId = o.Str("author_id");
        TimeText = o.Str("_time_text");
        Timestamp = o.Int("timestamp", -1);
        LikeCount = o.Int("like_count", -1);
        IsFavorited = o.Bool("is_favorited");
        AuthorIsUploader = o.Bool("author_is_uploader");
        IsPinned = o.Bool("is_pinned");
    }

    /* Thread yt-dlp's flat comment list.
     *
     * yt-dlp writes comments as a FLAT array with a `parent` field that is
     * either "root" or the parent's id, IN NO GUARANTEED ORDER -- a reply can
     * appear before its parent, so this is a two-pass job rather than a fold.
     *
     * Top level is pinned first, then most-liked: the same order YouTube itself
     * shows, which matters because a transcript of a comment section in
     * arbitrary order is a different document from the one people read. */
    public static List<Comment> Thread(List<JsonElement> raw)
    {
        var tops = new List<Comment>();
        var index = new Dictionary<string, int>(StringComparer.Ordinal);
        var orphans = new List<(string Parent, Comment Comment)>();

        foreach (var o in raw)
        {
            var parent = o.Str("parent");
            var c = new Comment(o);

            if (parent is null || parent == "root")
            {
                // First writer wins on a duplicate id: a later comment claiming
                // an id an earlier one already used must not silently steal its
                // replies.
                if (c.CommentId.Length > 0 && !index.ContainsKey(c.CommentId))
                    index[c.CommentId] = tops.Count;
                tops.Add(c);
            }
            else
            {
                orphans.Add((parent, c));
            }
        }

        foreach (var (parent, c) in orphans)
        {
            int slot = -1;
            if (index.TryGetValue(parent, out var found)) slot = found;
            else
            {
                /* yt-dlp's reply ids are "<parent>.<reply>", so the parent id is
                 * recoverable even when the parent field itself is unhelpful. */
                var dot = parent.IndexOf('.');
                if (dot > 0 && index.TryGetValue(parent[..dot], out var viaDot)) slot = viaDot;
            }

            if (slot >= 0)
            {
                tops[slot].Replies.Add(c);
            }
            else
            {
                /* A reply whose parent is genuinely absent -- a deleted comment,
                 * or a truncated fetch -- is shown at top level rather than
                 * dropped. Silently losing archived text would be the worse
                 * failure. */
                tops.Add(c);
            }
        }

        foreach (var c in tops)
        {
            // OrderBy is a stable sort in LINQ, so replies with equal (or
            // absent) timestamps keep the order yt-dlp wrote them in.
            var ordered = c.Replies.OrderBy(r => r.Timestamp < 0 ? 0 : r.Timestamp).ToList();
            c.Replies.Clear();
            c.Replies.AddRange(ordered);
        }

        /* Pinned first, then most-liked. List.Sort is an UNSTABLE introsort, so
         * the original position is carried as an explicit final tie-break --
         * without it the order of equally-liked comments, which is most of a
         * long thread, would differ between two reads of the same file.
         * (OrderBy would be stable and would not need the tie-break; the
         * explicit comparison is kept because it says out loud what is being
         * relied on.) */
        return tops
            .Select((c, offset) => (Comment: c, Offset: offset))
            .OrderByDescending(x => x.Comment.IsPinned)
            .ThenByDescending(x => Math.Max(x.Comment.LikeCount, 0))
            .ThenBy(x => x.Offset)
            .Select(x => x.Comment)
            .ToList();
    }

    /// This comment plus every reply under it.
    public static int TotalCount(List<Comment> tops) => tops.Sum(c => 1 + c.Replies.Count);
}

// MARK: - Transcript

public sealed class Cue
{
    public double Start { get; init; }
    public double End { get; set; }
    public string Text { get; init; } = "";
}

public static class Transcript
{
    /* A .vtt or .srt turned into a readable transcript.
     *
     * YouTube's auto-generated VTT is a ROLLING TWO-LINE DISPLAY: nearly every
     * cue repeats the previous cue's last line, and words carry inline karaoke
     * timestamps. Read as-is it is unusable as prose, so tags are stripped and
     * the repetition is collapsed. */
    public static List<Cue> Cues(string path)
    {
        byte[] bytes;
        try { bytes = File.ReadAllBytes(Paths.Extended(path)); }
        catch (Exception) { return new List<Cue>(); }

        var raw = Paths.DecodeText(bytes);

        /* Normalise line endings FIRST. A .vtt written on Windows -- which on
         * this platform is most of them -- would otherwise never match the
         * blank-line block separator, and the whole file would parse as one
         * block with no timeline. */
        var normalised = raw.Replace("\r\n", "\n").Replace('\r', '\n');

        var parsed = new List<Cue>();
        foreach (var block in normalised.Split("\n\n"))
        {
            var lines = block.Split('\n')
                             .Where(l => l.Trim().Length > 0)
                             .ToList();
            if (lines.Count == 0) continue;

            var timeIdx = lines.FindIndex(l => l.Contains("-->", StringComparison.Ordinal));
            if (timeIdx < 0) continue;

            var timeline = lines[timeIdx];
            var arrow = timeline.IndexOf("-->", StringComparison.Ordinal);
            if (arrow < 0) continue;

            var start = Seconds(timeline[..arrow]);
            if (start is null) continue;
            var end = Seconds(timeline[(arrow + 3)..]) ?? (start.Value + 3.0);

            var body = string.Join(" ", lines.Skip(timeIdx + 1));
            var text = CollapseWhitespace(UnescapeEntities(StripInlineTags(body)));
            if (text.Length == 0) continue;

            parsed.Add(new Cue { Start = start.Value, End = end, Text = text });
        }

        return CollapseRolling(parsed);
    }

    /* The rolling-display collapse.
     *
     * The comparison is against the previous cue's FULL text, not against what
     * was last emitted. That distinction is the whole algorithm:
     *
     *   cue 1  "the quick brown fox"
     *   cue 2  "the quick brown fox jumps over"
     *   cue 3  "the quick brown fox jumps over the lazy dog"
     *
     * Emitting the tail of cue 2 puts "jumps over" in the output. Comparing cue
     * 3 against THAT finds no common prefix, so cue 3 is emitted whole and the
     * duplication the collapse exists to remove comes straight back on the
     * third line. Keeping lastFull separate from the output is what makes the
     * third line "the lazy dog".
     *
     * The same defect was in orchid-cobalt's src-tauri/src/archive.rs and was
     * found by the C port's own test. It works for two lines and then silently
     * stops, which is why the test in this suite is three lines deep and not
     * two. */
    private static List<Cue> CollapseRolling(List<Cue> cues)
    {
        var output = new List<Cue>();
        string? lastFull = null;

        foreach (var cue in cues)
        {
            if (lastFull is { } previous && output.Count > 0)
            {
                if (string.Equals(cue.Text, previous, StringComparison.Ordinal))
                {
                    output[^1].End = Math.Max(output[^1].End, cue.End);
                    continue;
                }
                /* The 12-character floor keeps a genuinely repeated short line
                 * ("Yeah." then "Yeah. Right.") from being chopped into
                 * fragments. */
                if (cue.Text.StartsWith(previous, StringComparison.Ordinal) && previous.Length > 12)
                {
                    var tail = cue.Text[previous.Length..].Trim();
                    if (tail.Length == 0)
                    {
                        output[^1].End = Math.Max(output[^1].End, cue.End);
                    }
                    else
                    {
                        output.Add(new Cue { Start = cue.Start, End = cue.End, Text = tail });
                    }
                    lastFull = cue.Text;
                    continue;
                }
            }

            output.Add(cue);
            lastFull = cue.Text;
        }
        return output;
    }

    /* [hh:]mm:ss[.,]mmm -- the leading run of digits, colons and a decimal
     * mark. Anything after it (VTT cue settings such as "align:start
     * position:0%") is ignored. */
    public static double? Seconds(string text)
    {
        var trimmed = text.Trim();
        var head = new StringBuilder();
        foreach (var ch in trimmed)
        {
            if (char.IsAsciiDigit(ch) || ch == ':' || ch == '.' || ch == ',') head.Append(ch);
            else break;
        }
        if (head.Length == 0) return null;

        var headText = head.ToString();
        var whole = headText;
        var fraction = "";
        var mark = headText.IndexOfAny(new[] { '.', ',' });
        if (mark >= 0)
        {
            whole = headText[..mark];
            fraction = headText[(mark + 1)..];
        }

        // Not RemoveEmptyEntries: "::30" must fail rather than silently become
        // 30 seconds by having its empty fields dropped.
        var parts = whole.Split(':');
        double h = 0, m = 0, s = 0;
        if (parts.Length == 3)
        {
            h = ParseOrZero(parts[0]);
            m = ParseOrZero(parts[1]);
            s = ParseOrZero(parts[2]);
        }
        else if (parts.Length == 2)
        {
            m = ParseOrZero(parts[0]);
            s = ParseOrZero(parts[1]);
        }
        else
        {
            return null;
        }

        /* ".5" is 500ms, not 5ms -- pad on the RIGHT to three digits. */
        double ms = 0;
        if (fraction.Length > 0)
        {
            var padded = (fraction + "000")[..3];
            ms = ParseOrZero(padded);
        }

        return h * 3600 + m * 60 + s + ms / 1000;
    }

    private static double ParseOrZero(string s)
        => double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out var v) ? v : 0;

    /* Removes <c>, </c>, <v Name>, and the per-word <00:00:01.234> karaoke
     * timestamps YouTube's ASR emits. A hand-rolled scanner rather than a
     * regular expression: the grammar is "everything between < and >", and a
     * Regex over a multi-megabyte transcript is not free. */
    public static string StripInlineTags(string s)
    {
        var output = new StringBuilder(s.Length);
        var depth = 0;
        foreach (var ch in s)
        {
            if (ch == '<') depth++;
            else if (ch == '>') { if (depth > 0) depth--; }
            else if (depth == 0) output.Append(ch);
        }
        return output.ToString();
    }

    public static string UnescapeEntities(string s)
    {
        // &amp; is done LAST, so "&amp;lt;" comes out as "&lt;" and not as "<".
        // Doing it first turns an escaped entity into a real one, which is the
        // classic ordering bug in a hand-rolled unescaper.
        var output = s
            .Replace("&lt;", "<")
            .Replace("&gt;", ">")
            .Replace("&quot;", "\"")
            .Replace("&#39;", "'")
            .Replace("&nbsp;", " ")
            .Replace("&amp;", "&");
        return output;
    }

    public static string CollapseWhitespace(string s)
        => string.Join(" ", s.Split(new[] { ' ', '\t', '\n', '\r' },
                                    StringSplitOptions.RemoveEmptyEntries));

    /* The filenames cannot tell an auto-generated track from a human-written
     * one -- --write-subs and --write-auto-subs both land in Subtitles/ under
     * the same base name. The CONTENTS can: ASR output carries per-word karaoke
     * tags and cue-positioning directives that uploaded tracks do not.
     *
     * Only the head is read, and these files can be large. */
    public static bool IsAutoGenerated(string path)
    {
        try
        {
            using var stream = new FileStream(Paths.Extended(path), FileMode.Open,
                                              FileAccess.Read, FileShare.ReadWrite);
            var buffer = new byte[8000];
            var read = stream.Read(buffer, 0, buffer.Length);
            if (read <= 0) return false;

            var head = Paths.DecodeText(buffer[..read]);
            return head.Contains("<c.", StringComparison.Ordinal)
                || head.Contains("<c>", StringComparison.Ordinal)
                || head.Contains("align:start position:", StringComparison.Ordinal);
        }
        catch (Exception) { return false; }
    }
}
